File: agent/internal/repo/repo.go
// Package repo 提供工作区的只读文件浏览与 diff 查询,服务于 UI 的
// call/call-response 同步请求(文件树、读文件、变更列表、单文件 diff)。
// 全部操作强制限定在工作区目录内。
package repo

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
)

const (
	maxFileBytes = 1 << 20 // 单文件读取上限 1MB
	maxListItems = 2000
)

// Browser 一个工作区的只读浏览器。
type Browser struct {
	workdir string
}

// New 创建浏览器(workdir 须为绝对路径)。
func New(workdir string) *Browser { return &Browser{workdir: workdir} }

func (b *Browser) resolve(rel string) (string, error) {
	p := filepath.Join(b.workdir, rel)
	p = filepath.Clean(p)
	r, err := filepath.Rel(b.workdir, p)
	if err != nil || r == ".." || strings.HasPrefix(r, ".."+string(filepath.Separator)) {
		return "", fmt.Errorf("路径 %s 超出工作区", rel)
	}
	return p, nil
}

// Entry 目录项。
type Entry struct {
	Name  string `json:"name"`
	Path  string `json:"path"` // 相对工作区
	IsDir bool   `json:"is_dir"`
	Size  int64  `json:"size"`
}

// ListFiles 列出目录内容(单层,非递归)。dir 为空表示工作区根。
func (b *Browser) ListFiles(dir string) ([]Entry, error) {
	target, err := b.resolve(dir)
	if err != nil {
		return nil, err
	}
	items, err := os.ReadDir(target)
	if err != nil {
		return nil, err
	}
	var out []Entry
	for _, it := range items {
		if it.Name() == ".git" {
			continue
		}
		info, err := it.Info()
		if err != nil {
			continue
		}
		rel, _ := filepath.Rel(b.workdir, filepath.Join(target, it.Name()))
		out = append(out, Entry{
			Name: it.Name(), Path: filepath.ToSlash(rel),
			IsDir: it.IsDir(), Size: info.Size(),
		})
		if len(out) >= maxListItems {
			break
		}
	}
	// 目录在前,再按名排序
	sort.Slice(out, func(i, j int) bool {
		if out[i].IsDir != out[j].IsDir {
			return out[i].IsDir
		}
		return out[i].Name < out[j].Name
	})
	return out, nil
}

// ReadFile 读取文件内容(带上限)。
func (b *Browser) ReadFile(rel string) (string, error) {
	p, err := b.resolve(rel)
	if err != nil {
		return "", err
	}
	info, err := os.Stat(p)
	if err != nil {
		return "", err
	}
	if info.IsDir() {
		return "", fmt.Errorf("%s 是目录", rel)
	}
	if info.Size() > maxFileBytes {
		return "", fmt.Errorf("文件过大(%d 字节),超过 %d 上限", info.Size(), maxFileBytes)
	}
	data, err := os.ReadFile(p)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// Change 变更文件条目。Status: A(新增)/M(修改)/D(删除)。
type Change struct {
	Path   string `json:"path"`
	Status string `json:"status"`
}

// FileChanges 相对 HEAD 的变更列表(含未跟踪文件)。非 git 仓库返回空。
// 路径统一为相对 workdir:porcelain 输出的是仓库根相对路径,workdir 位于
// 仓库子目录时直接拿去当 pathspec 会永远匹配不上(diff 恒空);
// quotepath 关闭,否则非 ASCII 文件名(中文)被转成八进制转义乱码。
func (b *Browser) FileChanges() ([]Change, error) {
	if !b.isGitRepo() {
		return nil, nil
	}
	prefix, _ := b.git("rev-parse", "--show-prefix")
	prefix = strings.TrimSpace(prefix)
	// pathspec "." 限定 workdir 子树:子目录会话不该看到仓库其他地方的改动
	out, _ := b.git("-c", "core.quotepath=false", "status", "--porcelain=v1", "--untracked-files=all", "--", ".")
	var changes []Change
	for line := range strings.SplitSeq(out, "\n") {
		if len(line) < 4 {
			continue
		}
		code := strings.TrimSpace(line[:2])
		path := strings.TrimSpace(line[3:])
		// 处理重命名 "old -> new"
		if i := strings.Index(path, " -> "); i >= 0 {
			path = path[i+4:]
		}
		path = strings.Trim(path, `"`)
		// 仓库根相对 → workdir 相对(前缀之外的条目丢弃,双保险)
		if prefix != "" {
			if !strings.HasPrefix(path, prefix) {
				continue
			}
			path = strings.TrimPrefix(path, prefix)
		}
		status := "M"
		switch {
		case strings.Contains(code, "?"), strings.Contains(code, "A"):
			status = "A"
		case strings.Contains(code, "D"):
			status = "D"
		}
		changes = append(changes, Change{Path: path, Status: status})
	}
	sort.Slice(changes, func(i, j int) bool { return changes[i].Path < changes[j].Path })
	return changes, nil
}

// FileDiff 单个文件相对 HEAD 的 unified diff。未跟踪文件构造为全新增 diff。
func (b *Browser) FileDiff(rel string) (string, error) {
	if _, err := b.resolve(rel); err != nil {
		return "", err
	}
	if !b.isGitRepo() {
		return "", fmt.Errorf("非 git 仓库,无法生成 diff")
	}
	// 已跟踪文件:直接 diff HEAD(rel 为 workdir 相对,与 FileChanges 一致)
	out, err := b.git("-c", "core.quotepath=false", "diff", "HEAD", "--", rel)
	if err == nil && strings.TrimSpace(out) != "" {
		return out, nil
	}
	// 未跟踪文件:git diff --no-index 生成新增 diff
	if untracked, _ := b.git("ls-files", "--others", "--exclude-standard", "--", rel); strings.TrimSpace(untracked) != "" {
		d, _ := b.gitAllowFail("-c", "core.quotepath=false", "diff", "--no-index", "--", os.DevNull, rel)
		return d, nil
	}
	return out, nil
}

func (b *Browser) isGitRepo() bool {
	out, err := b.git("rev-parse", "--is-inside-work-tree")
	return err == nil && strings.TrimSpace(out) == "true"
}

func (b *Browser) git(args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = b.workdir
	out, err := cmd.Output()
	return string(out), err
}

// gitAllowFail 用于 diff --no-index(有差异时退出码为 1,非错误)。
func (b *Browser) gitAllowFail(args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = b.workdir
	out, _ := cmd.Output()
	return string(out), nil
}

File: desktop/src/repo.rs
// 工作区只读文件浏览与 diff 查询(agent/internal/repo 的 Rust 移植)。
// 应答字段与内核 WS call 完全对齐:{result} / {error},UI 归约层零改动。
// 全部操作强制限定在工作区目录内。
//
// WSL 模式(kernel_env = "wsl:<发行版>"):workdir 是 guest 内 Linux 路径,
// 文件系统操作经 \\wsl$\<发行版> UNC 访问,git 经 wsl.exe 在 guest 内执行
// (UNC 上跑 Windows git 会撞 ownership 校验且行尾语义不对)。

use std::path::{Path, PathBuf};
use std::process::Command;

use serde_json::{json, Value};

const MAX_FILE_BYTES: u64 = 1 << 20; // 单文件读取上限 1MB
const MAX_LIST_ITEMS: usize = 2000;

/// 一次 repo 查询的执行环境。
pub struct RepoCtx {
    /// 工作区绝对路径(WSL 模式下是 guest 内 Linux 路径)
    pub workdir: String,
    /// WSL 发行版(仅 Windows + kernel_env=wsl:* 时 Some)
    pub wsl_distro: Option<String>,
}

impl RepoCtx {
    /// 本地文件系统视角的工作区根(WSL 模式转 UNC)。
    fn fs_root(&self) -> PathBuf {
        match &self.wsl_distro {
            Some(d) => crate::wsl::unc_path(d, &self.workdir),
            None => PathBuf::from(&self.workdir),
        }
    }

    /// 解析相对路径并防目录穿越;返回本地 fs 视角的绝对路径。
    fn resolve(&self, rel: &str) -> Result<PathBuf, String> {
        // 归一化组件级校验:拒绝绝对路径与任何 .. 成分(简单可靠,无需 canonicalize
        // ——目标文件可能尚不存在,canonicalize 会失败)
        let p = Path::new(rel);
        if p.is_absolute()
            || p.components().any(|c| {
                matches!(c, std::path::Component::ParentDir | std::path::Component::Prefix(_))
            })
        {
            return Err(format!("路径 {rel} 超出工作区"));
        }
        Ok(self.fs_root().join(rel))
    }

    /// 在工作区内执行 git 的唯一通道(WSL 模式经 wsl.exe 在 guest 内跑)。
    /// allow_fail:非零退出码不视为错误,只取 stdout(diff --no-index 有
    /// 差异时退出码为 1 这类"正常失败")。
    fn run_git(&self, args: &[&str], allow_fail: bool) -> Result<String, String> {
        let mut cmd = match &self.wsl_distro {
            Some(d) => {
                let mut c = Command::new(crate::wsl::wsl_exe());
                c.args(["-d", d, "--cd", &self.workdir, "--exec", "git"]).args(args);
                c
            }
            None => {
                let mut c = Command::new("git");
                c.current_dir(&self.workdir).args(args);
                c
            }
        };
        crate::wsl::no_console(&mut cmd); // Windows 下每次文件树/diff 查询不闪黑窗
        let out = cmd.output().map_err(|e| format!("git 执行失败: {e}"))?;
        if !allow_fail && !out.status.success() {
            return Err(String::from_utf8_lossy(&out.stderr).trim().to_string());
        }
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }

    fn git(&self, args: &[&str]) -> Result<String, String> {
        self.run_git(args, false)
    }

    fn git_allow_fail(&self, args: &[&str]) -> String {
        // spawn 失败也吞掉返回空:调用方(未跟踪文件 diff)按无差异降级
        self.run_git(args, true).unwrap_or_default()
    }

    fn is_git_repo(&self) -> bool {
        self.git(&["rev-parse", "--is-inside-work-tree"])
            .map(|s| s.trim() == "true")
            .unwrap_or(false)
    }
}

/// 统一入口:kind 分派,返回 {result} 或 {error}(与内核 call-response 载荷同构)。
pub fn dispatch(ctx: &RepoCtx, kind: &str, payload: &Value) -> Value {
    let path = payload.get("path").and_then(|v| v.as_str()).unwrap_or("");
    let r = match kind {
        "repo_file_list" => list_files(ctx, path),
        "repo_read_file" => read_file(ctx, path),
        "repo_file_changes" => file_changes(ctx),
        "repo_file_diff" => file_diff(ctx, path),
        "repo_reveal" => reveal(ctx, path),
        _ => Err(format!("未知 call kind: {kind}")),
    };
    match r {
        Ok(v) => json!({ "result": v }),
        Err(e) => json!({ "error": e }),
    }
}

/// 列出目录内容(单层,非递归)。dir 为空表示工作区根。目录在前,再按名排序。
fn list_files(ctx: &RepoCtx, dir: &str) -> Result<Value, String> {
    let target = ctx.resolve(dir)?;
    let items = std::fs::read_dir(&target).map_err(|e| format!("读取目录失败: {e}"))?;
    let mut out: Vec<(bool, String, u64)> = Vec::new();
    for it in items.flatten() {
        let name = it.file_name().to_string_lossy().into_owned();
        if name == ".git" {
            continue;
        }
        let Ok(md) = it.metadata() else { continue };
        out.push((md.is_dir(), name, md.len()));
        if out.len() >= MAX_LIST_ITEMS {
            break;
        }
    }
    out.sort_by(|a, b| b.0.cmp(&a.0).then(a.1.cmp(&b.1)));
    let base = if dir.is_empty() { String::new() } else { format!("{}/", dir.trim_end_matches('/')) };
    let entries: Vec<Value> = out
        .into_iter()
        .map(|(is_dir, name, size)| {
            json!({ "name": name, "path": format!("{base}{name}"), "is_dir": is_dir, "size": size })
        })
        .collect();
    Ok(Value::Array(entries))
}

/// 读取文件内容(1MB 上限)。
fn read_file(ctx: &RepoCtx, rel: &str) -> Result<Value, String> {
    let p = ctx.resolve(rel)?;
    let md = std::fs::metadata(&p).map_err(|e| format!("读取失败: {e}"))?;
    if md.is_dir() {
        return Err(format!("{rel} 是目录"));
    }
    if md.len() > MAX_FILE_BYTES {
        return Err(format!("文件过大({} 字节),超过 {} 上限", md.len(), MAX_FILE_BYTES));
    }
    let data = std::fs::read(&p).map_err(|e| format!("读取失败: {e}"))?;
    Ok(json!({ "path": rel, "content": String::from_utf8_lossy(&data) }))
}

/// 相对 HEAD 的变更列表(含未跟踪文件)。非 git 仓库返回空。
/// 路径统一为相对 workdir(porcelain 输出仓库根相对路径,须剥前缀);
/// quotepath 关闭,否则非 ASCII 文件名被转成八进制转义乱码。
fn file_changes(ctx: &RepoCtx) -> Result<Value, String> {
    if !ctx.is_git_repo() {
        return Ok(Value::Array(vec![]));
    }
    let prefix = ctx.git(&["rev-parse", "--show-prefix"]).unwrap_or_default().trim().to_string();
    let out = ctx
        .git(&["-c", "core.quotepath=false", "status", "--porcelain=v1", "--untracked-files=all", "--", "."])
        .unwrap_or_default();
    let mut changes: Vec<(String, &'static str)> = Vec::new();
    for line in out.lines() {
        if line.len() < 4 {
            continue;
        }
        let code = line[..2].trim();
        let mut path = line[3..].trim().to_string();
        // 处理重命名 "old -> new"
        if let Some(i) = path.find(" -> ") {
            path = path[i + 4..].to_string();
        }
        path = path.trim_matches('"').to_string();
        // 仓库根相对 → workdir 相对(前缀之外的条目丢弃,双保险)
        if !prefix.is_empty() {
            match path.strip_prefix(&prefix) {
                Some(p) => path = p.to_string(),
                None => continue,
            }
        }
        let status = if code.contains('?') || code.contains('A') {
            "A"
        } else if code.contains('D') {
            "D"
        } else {
            "M"
        };
        changes.push((path, status));
    }
    changes.sort_by(|a, b| a.0.cmp(&b.0));
    Ok(Value::Array(
        changes.into_iter().map(|(p, s)| json!({ "path": p, "status": s })).collect(),
    ))
}

/// 单个文件相对 HEAD 的 unified diff。未跟踪文件构造为全新增 diff。
fn file_diff(ctx: &RepoCtx, rel: &str) -> Result<Value, String> {
    ctx.resolve(rel)?;
    if !ctx.is_git_repo() {
        return Err("非 git 仓库,无法生成 diff".into());
    }
    // 已跟踪文件:直接 diff HEAD(rel 为 workdir 相对,与 file_changes 一致)
    if let Ok(out) = ctx.git(&["-c", "core.quotepath=false", "diff", "HEAD", "--", rel]) {
        if !out.trim().is_empty() {
            return Ok(json!({ "path": rel, "diff": out }));
        }
    }
    // 未跟踪文件:git diff --no-index 生成新增 diff
    let untracked = ctx.git(&["ls-files", "--others", "--exclude-standard", "--", rel]).unwrap_or_default();
    if !untracked.trim().is_empty() {
        // guest 内跑 git 时 null 设备始终是 /dev/null;仅本机 Windows 用 NUL
        let null_dev = if ctx.wsl_distro.is_none() && cfg!(windows) { "NUL" } else { "/dev/null" };
        let d = ctx.git_allow_fail(&["-c", "core.quotepath=false", "diff", "--no-index", "--", null_dev, rel]);
        return Ok(json!({ "path": rel, "diff": d }));
    }
    Ok(json!({ "path": rel, "diff": "" }))
}

/// 在系统文件管理器中定位路径:目录直接打开,文件在父目录中选中。
/// WSL 模式路径已经是 UNC,explorer 直接可开。
fn reveal(ctx: &RepoCtx, rel: &str) -> Result<Value, String> {
    let p = ctx.resolve(rel)?;
    let md = std::fs::metadata(&p).map_err(|e| format!("路径不存在: {e}"))?;
    let r = if cfg!(target_os = "macos") {
        if md.is_dir() {
            Command::new("open").arg(&p).spawn()
        } else {
            Command::new("open").arg("-R").arg(&p).spawn()
        }
    } else if cfg!(windows) {
        if md.is_dir() {
            Command::new("explorer").arg(&p).spawn()
        } else {
            Command::new("explorer").arg(format!("/select,{}", p.display())).spawn()
        }
    } else {
        let dir = if md.is_dir() { p.clone() } else { p.parent().unwrap_or(&p).to_path_buf() };
        Command::new("xdg-open").arg(&dir).spawn()
    };
    // Start 不等退出:explorer.exe 成功也返回非零码,等待会误报错误
    r.map_err(|e| format!("打开文件管理器失败: {e}"))?;
    Ok(json!({ "ok": true }))
}
